feat: 글 신고 CRD 구현

aegis1920 opened the pull request:
**이슈 번호**

resolved: #14 

**작업 내용**

1. 글 신고 CRD
2. C는 사용자만, RD는 관리자만 가능
3. 글 신고 전체 조회할 때 Page 별로 조회하도록 구현
4. 글 신고 삭제는 soft delete 하지 않음 (논의 필요)
5. post_report는 post, user를 가지고 있고 post와 user는 post_report와 관계를 맺고 있지 않음

**테스트 작성 여부**

- [X] Test case

aegis1920 committed:
feat: 글 신고 인수테스트 작성 및 컨트롤러 Mock 데이터 반환

aegis1920 committed:
feat: 글 신고 작성 서비스 테스트 및 도메인 작성

- V5__Issue_14_Create_PostReport SQL 작성
- PostReportAssembler 작성
- PostReportCreateRequest DTO 작성
- PostReportResponse DTO 작성
- init-table에 PostReport 삭제

aegis1920 committed:
feat: 글 신고 작성 서비스 및 컨트롤러 연동

aegis1920 committed:
feat: 관리자 글 신고 조회 인수테스트 및 컨트롤러 Mock 데이터 반환

aegis1920 committed:
feat: 관리자 글 신고 조회 서비스 구현 및 테스트 작성

- 글 신고 서비스와 컨트롤러와도 연동

aegis1920 committed:
feat: 관리자 페이지 별 글 신고 조회 서비스 구현 및 테스트 작성

aegis1920 committed:
feat: 관리자 글 신고 삭제 서비스 및 테스트 구현

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/dto/PostReportResponse.java:
```diff
@@ -0,0 +1,41 @@
+    private UserResponse creator;
+    private LocalDateTime createdAt;
+    private LocalDateTime modifiedAt;
+
+
```
공백 한 줄 없어도 될것같네요~

aegis1920 replied:
엇 감사합니다 :)

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/PostService.java:
```diff
@@ -133,4 +139,16 @@ public void pressZzang(Long postId) {
         Post post = findPostBy(postId);
         post.pressZzang(user);
     }
+
+    @Transactional
+    public PostReportResponse createPostReport(Long postId,
+        PostReportCreateRequest postReportCreateRequest) {
+        User user = (User) authenticationFacade.getPrincipal();
+        Post post = findPostBy(postId);
+
+        PostReport postReport = PostReportAssembler.assemble(postReportCreateRequest, post, user);
+
+        PostReport savedPostReport = postReportRepository.save(postReport);
+        return PostReportResponse.of(savedPostReport);
+    }
```
Report Service 가 있는데 createPostReport 는 PostService 에서 하는 이유가 있나요?!

ReportService 도 PostReportRepository 에 의존하고
PostService 도 PostReportRepository에 의존하다보니
객체의 책임 분리가 모호해진것이 아닌가 싶은 생각이 들었습니다!

`Post post = findPostBy(postId);` 이부분 때문이라면
PostService 가 PostReportRepository 에 의존하는 것 보다는 ReportService에 의존하는 것은 어떨까요?!

aegis1920 replied:
처음에 설계를 할 때 Create는 유저만, Read, Delete는 관리자만 가능하다고 생각했습니다!
그래서 Create 기능은 유저가 들어갈 수 있도록 PostController에서 구현했고
 Read, Delete는 관리자만 들어갈 수 있도록 따로 ReportAdminController를 만들어 구현했습니다.

PostController에 PostService가 있으니 자연스럽게 Create는 PostService에서 하게 됐는데 Create를 ReportService에서 한다면 PostController에 PostService와 ReportService가 있게 됩니다.

-> 하나의 컨트롤러에 두 개의 서비스를 의존하고 있어도 상관없나요? 어떻게 생각하시나욤?

yelimkim98 replied:
흐음 그렇군요..
제 의견 대로라면 ReportAdminController 대신 ReportController가 있어서 CRD 를 다 여기서 처리하는 구조여야 겠네요,,
관점의 차이라고 느껴지는 부분이네요...!!

아얘 이렇게 구조를 바꾸는것도 괜찮다고 생각하고,
아니면 PostService 가 ReportService 에 의존하는 것은 어떤가요??
컨트롤러가 두개의 서비스를 쓰는게 아니라
서비스가 서비스에게 의존하는 것이요..!
(이것도 최선이라는 생각은 들지 않지만요..)

사실 지금대로도 크게 상관은 없습니다.. ㅎㅎ (그래서 approve를 한 것이구요 ㅎ)

aegis1920 replied:
사용자용 ReportController를 하나 더 만들었습니다! createPostReport는 ReportService에 구현해놓았고, ReportService는 PostRepository만 의존하게 했습니다.
가져오는 메서드 하나를 사용하자고 PostService를 의존하게 하는 건 너무 큰 것 같아서요 :)

yelimkim98 reviewed (approved):
수고하셨습니다!! 👍 💯

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/PostController.java:
```diff
@@ -84,4 +86,14 @@
+    @PostMapping("/{postId}/reports")
+    public ResponseEntity<Void> createPostReport(@PathVariable Long postId, @RequestBody
+        PostReportCreateRequest postReportCreateRequest) {
+        final PostReportResponse postReportResponse = postService
```
다른 곳에도 final을 붙이지 않고 있어서, final을 붙이지 않아도 될 것 같아요! (암묵적..)

aegis1920 replied:
자동완성하면서 붙여졌던 것 같네요! 수정하겠습니다!

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/PostController.java:
```diff
@@ -84,4 +86,14 @@
+        PostReportCreateRequest postReportCreateRequest) {
+        final PostReportResponse postReportResponse = postService
+            .createPostReport(postId, postReportCreateRequest);
+        return ResponseEntity
```
개행 한 칸 띄워줘도 좋을 것 같아요~!

aegis1920 replied:
넵넵!

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/post/PostReport.java:
```diff
@@ -0,0 +1,51 @@
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "creator_id", nullable = false)
+    private User creator;
```
신고자라는 의미로 `reporter`도 괜찮을 것 같네요.

Hamliet replied:
동의합니다!

aegis1920 replied:
오 좀 더 명확할 것 같아요 👍

jamie9504 reviewed (commented):
도메인 관련 코멘트 남겨두어야할 것 같아서 빠르게 남겨둬봅니다!

 신고의 경우, 글이 계속 변경될 수 있는데(예 - 광고글이었다가, 다시 일반 글로 변경), 그 신고 시점의 snapshot을 뜨던지, 사용자에게 screenshot 등을 요구하던지의 필요성이 느껴지는데요!
 어떤 것을 선택하느냐에 따라 요청방법과 필드가 달라질 수 있겠네요, 빙봉은 혹시 어떤 방식이 더 좋다고 생각하시나요?

 저는 조금 고민해봤는데 snapshot을 뜨면, 데이터 용량은 늘어날 수 있을 것 같아요.
    > postImage까지 모두 데이터 카피를 해야하는데, postImage의 필드가 바뀔 수도 있겠네요.
 screenshot을 뜨자니 사용자 입장에서 좀 번거로와서 신고를 꺼릴 수도 있을 것 같구요.

:) 빙봉의 의견이 궁금하네요.

aegis1920 committed:
Merge branch 'develop' into feature/#14

Hamliet commented on back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/acceptance/UserAcceptanceTest.java:
```diff
@@ -41,14 +41,14 @@
+    @MockBean
+    private MailAuthRepository mailAuthRepository;
+
     @BeforeEach
     void setUp() {
         RestAssured.port = port;
     }
 
-    @MockBean
-    private MailAuthRepository mailAuthRepository;
-
```
👍

Hamliet reviewed (approved):
빙봉 역시 깔끔하게 잘 구현하셨네요! :)
다른 분들 피드백 이외에 추가로 리뷰할 부분 없어 바로 Approve 하겠습니다.
수고하셨습니다!

aegis1920 committed:
feat: ReportController 추가 및 createPostReport 메서드를 ReportService로 옮김

- ReportService 가 PostRepository 를 가지고 있음

aegis1920 committed:
feat: PostReport의 creator 필드 reporter 로 수정 및 외래키 제약조건 이름 변경

jamie9504 reviewed (approved):
Approve~!

aegis1920 committed:
feat: 신고할 때 해당 Post의 스냅샷을 저장하는 기능 추가 및 테스트 수정

- PostSnapshot 도메인 추가
- PostReport가 Post를 저장하는 게 아니라 PostSnapshot을 저장
- 조인 칼럼의 제약조건에 이름 추가
- init-table에 POST_REPORT_POST_IMAGE_URLS 테이블 삭제

aegis1920 committed:
feat: 이미지가 있는 글 신고 테스트 추가

aegis1920 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#14

aegis1920 committed:
chore: 마이그레이션 번호 수정

aegis1920 committed:
chore: 마이그레이션 번호 수정 및 UserResponse가 아닌 UserSimpleResponse 객체를 가져오도록 수정

aegis1920 committed:
Merge branch 'feature/#14' of https://github.com/aegis1920/2020-legeno-around-here into feature/#14
